lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -0,0 +1,592 @@
+    auto outF = [&](unsigned idx) { return op.getResName(idx).str(); };
+    llvm::transform(
+        llvm::enumerate(op.getType().getInputs()), std::back_inserter(inputs),
+        [&](auto it) { return builder.getStringAttr(inF(it.index())); });
+    llvm::transform(
+        llvm::enumerate(op.getType().getResults()), std::back_inserter(outputs),
+        [&](auto it) { return builder.getStringAttr(outF(it.index())); });
+  }
+
+  Builder builder;
+  llvm::SmallVector<StringAttr> inputs;
+  llvm::SmallVector<StringAttr> outputs;
+};
+
+/// Construct a name for creating HW sub-module.
+static std::string getSubModuleName(Operation *oldOp) {
+  if (auto instanceOp = dyn_cast<handshake::InstanceOp>(oldOp); instanceOp)
+    return instanceOp.getModule().str();
+
+  std::string subModuleName = getBareSubModuleName(oldOp);
+
+  // Add value of the constant operation.
+  if (auto constOp = dyn_cast<handshake::ConstantOp>(oldOp)) {
+    if (auto intAttr = constOp.getValue().dyn_cast<IntegerAttr>()) {
+      auto intType = intAttr.getType();
+
+      if (intType.isSignedInteger())
+        subModuleName += "_c" + std::to_string(intAttr.getSInt());
+      else if (intType.isUnsignedInteger())
+        subModuleName += "_c" + std::to_string(intAttr.getUInt());
+      else
+        subModuleName += "_c" + std::to_string((uint64_t)intAttr.getInt());
+    } else
+      oldOp->emitError("unsupported constant type");
+  }
+
+  // Add discriminating in- and output types.
+  auto [inTypes, outTypes] = getHandshakeDiscriminatingTypes(oldOp);
+  if (!inTypes.empty())
+    subModuleName += "_in";
+  for (auto inType : inTypes)
+    subModuleName += getTypeName(oldOp->getLoc(), inType);
+
+  if (!outTypes.empty())
+    subModuleName += "_out";
+  for (auto outType : outTypes)
+    subModuleName += getTypeName(oldOp->getLoc(), outType);
+
+  // Add memory ID.
+  if (auto memOp = dyn_cast<handshake::MemoryOp>(oldOp))
+    subModuleName += "_id" + std::to_string(memOp.id());
+
+  // Add compare kind.
+  if (auto comOp = dyn_cast<mlir::arith::CmpIOp>(oldOp))
+    subModuleName += "_" + stringifyEnum(comOp.getPredicate()).str();
+
+  // Add buffer information.
+  if (auto bufferOp = dyn_cast<handshake::BufferOp>(oldOp)) {
+    subModuleName += "_" + std::to_string(bufferOp.getNumSlots()) + "slots";
+    if (bufferOp.isSequential())
+      subModuleName += "_seq";
+    else
+      subModuleName += "_fifo";
+  }
+
+  // Add control information.
+  if (auto ctrlInterface = dyn_cast<handshake::ControlInterface>(oldOp);
+      ctrlInterface && ctrlInterface.isControl()) {
+    // Add some additional discriminating info for non-typed operations.
+    subModuleName += "_" + std::to_string(oldOp->getNumOperands()) + "ins_" +
+                     std::to_string(oldOp->getNumResults()) + "outs";
+    subModuleName += "_ctrl";
+  } else {
+    assert((!inTypes.empty() || !outTypes.empty()) &&
+           "Non-control operators must provide discriminating type info");
+  }
```

Comment:
Is this a precondition, or a case that you haven't implemented yet?

Reply:
This is made an assert due to the above logic ideally being able to capture all possible permutations of a given operation. If not, there's a problem in the logic!

Reply:
Ok, that's not obvious, since the assert makes it seem like something that the definition of the operator should have provided, rather than representing an internal invariant of this code.  Maybe "Discriminating type info not found for non-control operator"?